### DISCUSSION TITLE
does this work with opencv?

### DISCUSSION THREAD

**Discussion by Raj2032** [robertknight/ocrs#69] [General]
can this work with opencv, the idea is that have a camera and in real time it can detect words by in real time covnerting words into text?

**Comment by robertknight** [comment]
I haven't tried using this with OpenCV, but in principle it should be possible. Inference speed will depend a lot on the hardware you run it on. On my 4-year old Intel laptop it takes about 1s end-to-end for a typical image. If using the Rust library you can run just the text detection, which will be quicker (~200ms) and then run recognition only if something is seen.

**Comment by Raj2032** [comment]
> On my 4-year old Intel laptop it takes about 1s end-to-end for a typical image

That's quite impressive :)

> If using the Rust library you can run just the text detection step more quickly (~200ms) and then run recognition only if something is seen.

Sorry I didn't quite understand, are you saying don't run it as a software, but instead use it as a library and incorporate it into my opencv code?

**Comment by robertknight** [comment]
> Sorry I didn't quite understand, are you saying don't run it as a software, but instead use it as a library and incorporate it into my opencv code?

You can do either. The command-line tool has the advantage you can call it from any tool which can run a program. The library gives you more flexibility to eg. run only detection, or process a batch of images at once etc.

**Comment by Raj2032** [comment]
@robertknight Hey mate, so what is meant by "run only detection"?  What else does the program do?

**Comment by robertknight** [comment]
If you run the `ocrs` program normally on an image, the output is the text that was found. Internally it uses a three-step process where it first detects which pixels in the input are text, then attempts to determine how the text pixels should be grouped into words and lines, then attempts to read the text from each line. If you are using the library, you can run each of these steps independently. So if you only need to know whether an image contains text, but not what the text is, you can run just the first step.

**Reply by Raj2032** [reply]
I see now thank you mate :)

**Comment by zanaviska** [comment]
So I decided to try. And started with modifying [example](https://github.com/robertknight/ocrs/blob/71ba35e7c863d9c1d0b5676e094576677bb0f37b/ocrs/examples/hello_ocr.rs#L46). It ended up with the same code, but `ImageSource` I now generate like this
```
let img_source = ocrs::ImageSource::from_bytes(
        unsafe {
            std::slice::from_raw_parts(
                img_cv.data(),
                (img_cv.cols() * img_cv.rows() * img_cv.channels()) as usize,
            )
        },
        (img_cv.cols() as u32, img_cv.rows() as u32),
    )
    .unwrap();
```

I do not know why, but results(e.g. text position) is never absolutely the same as with example approach. It always differ by a few pixels. About text recognition. It recgnizes existing text as good as example version does. But when recognizing unexisting text(this happened to me for both example and opencv), model generates different results
